pkg/handlers/generic/lifecycle/cni/cilium/handler.go

```diff
@@ -231,11 +239,129 @@ func (c *CiliumCNI) apply(
 		return
 	}
 
-	if err := strategy.Apply(ctx, cluster, targetNamespace, log); err != nil {
+	if err := runApply(ctx, c.client, cluster, strategy, targetNamespace, log); err != nil {
 		resp.SetStatus(runtimehooksv1.ResponseStatusFailure)
 		resp.SetMessage(err.Error())
 		return
 	}
 
 	resp.SetStatus(runtimehooksv1.ResponseStatusSuccess)
 }
+
+func runApply(
+	ctx context.Context,
+	client ctrlclient.Client,
+	cluster *clusterv1.Cluster,
+	strategy addons.Applier,
+	targetNamespace string,
+	log logr.Logger,
+) error {
+	if err := strategy.Apply(ctx, cluster, targetNamespace, log); err != nil {
+		return err
+	}
+
+	// If skip kube-proxy is not set, return early.
+	// Otherwise, wait for Cilium to be rolled out and then cleanup kube-proxy if installed.
+	if !capiutils.SkipKubeProxy(cluster) {
+		return nil
+	}
+
+	log.Info(
+		fmt.Sprintf("Waiting for Cilium to be ready for cluster %s", ctrlclient.ObjectKeyFromObject(cluster)),
+	)
+	if err := waitForCiliumToBeReady(ctx, client, cluster); err != nil {
+		return fmt.Errorf("failed to wait for Cilium to be ready: %w", err)
+	}
+
+	log.Info(
+		fmt.Sprintf("Cleaning up kube-proxy for cluster %s", ctrlclient.ObjectKeyFromObject(cluster)),
+	)
+	if err := cleanupKubeProxy(ctx, client, cluster); err != nil {
+		return fmt.Errorf("failed to cleanup kube-proxy: %w", err)
+	}
+
+	return nil
+}
+
+const (
+	kubeProxyName      = "kube-proxy"
+	kubeProxyNamespace = "kube-system"
+)
+
+func waitForCiliumToBeReady(
```

Comment:
Initially I tried to just set HCP's helm `wait` option, but during upgrades it was getting `Ready` prematurely.
Keeping the wait outside also makes it possible to support the wait for other providers besides helm.